File: Basic/25.Solve the Activity Selection Problem/Solution.java
import java.util.*;

class Activity {
    int start, end;

    Activity(int start, int end) {
        this.start = start;
        this.end = end;
    }
}

public class ActivitySelection {

    public static void activitySelection(List<Activity> activities) {
        activities.sort(Comparator.comparingInt(a -> a.end));

        System.out.println("Selected activities (start, end):");
        int lastEnd = 0;
        for (Activity activity : activities) {
            if (activity.start >= lastEnd) {
                System.out.println("(" + activity.start + ", " + activity.end + ")");
                lastEnd = activity.end;
            }
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of activities: ");
        int n = sc.nextInt();

        List<Activity> activities = new ArrayList<>();
        System.out.println("Enter start and end times of activities:");
        for (int i = 0; i < n; i++) {
            System.out.print("Activity " + (i + 1) + ": ");
            int start = sc.nextInt();
            int end = sc.nextInt();
            activities.add(new Activity(start, end));
        }

        activitySelection(activities);
    }
}

File: Basic/25.Solve the Activity Selection Problem/Solution.py
def activity_selection(activities):
    # Sort activities based on their end time
    activities.sort(key=lambda x: x[1])

    print("Selected activities (start, end):")
    last_end = 0
    for start, end in activities:
        if start >= last_end:
            print(f"({start}, {end})")
            last_end = end

if __name__ == "__main__":
    n = int(input("Enter the number of activities: "))
    activities = []

    print("Enter start and end times of activities:")
    for i in range(n):
        start, end = map(int, input(f"Activity {i + 1}: ").split())
        activities.append((start, end))

    activity_selection(activities)
